src/read_pair_iter.rs

```diff
@@ -267,42 +287,59 @@ impl ReadPairIter {
         i1: Option<P>,
         i2: Option<P>,
         r1_interleaved: bool,
+    ) -> Result<ReadPairIter, FastqError> {
+        Self::_new(
+            [
+                r1.as_ref().map(P::as_ref),
+                r2.as_ref().map(P::as_ref),
+                i1.as_ref().map(P::as_ref),
+                i2.as_ref().map(P::as_ref),
+            ],
+            r1_interleaved,
+        )
+    }
+
+    fn _new(
+        in_paths: [Option<&Path>; 4],
+        r1_interleaved: bool,
     ) -> Result<ReadPairIter, FastqError> {
         let mut iters = [None, None, None, None];
         let mut paths = [None, None, None, None];
 
-        for (idx, r) in [r1, r2, i1, i2].iter().enumerate() {
-            if let Some(ref p) = *r {
+        for (idx, r) in IntoIterator::into_iter(in_paths).enumerate() {
```

Comment:
```suggestion
        for (idx, r) in in_paths.into_iter().enumerate() {
```

Reply:
Same deal.  In 2018 edition, that would have a different semantic meaning.